mlir/test/Dialect/Affine/affine-loop-interchange.mlir

```diff
@@ -0,0 +1,344 @@
+// CHECK-NEXT:   for %[[IV1.*]] = 0 to 2048 {
+// CHECK-NEXT:      affine.store %{{.*}}, %{{.*}}[%[[IV0]], %[[IV1]]]
+// CHECK-NEXT:   }
+// CHECK-NEXT: }
+
+
+// -----
+
+// Test for interchange on multilevel imperfect nests.
+
+func @multilevel_imperfect_nest(%A: memref<2048x2048xf64>) {
+  %c0 = constant 0.0 : f64
+  %c1 = constant 1.0 : f64
+  affine.for %i = 0 to 2048 {
+    affine.for %j = 0 to 1024 {
+      affine.store %c0, %A[%j, %i] : memref<2048x2048xf64>
+    }
+    affine.for %j = 1024 to 2048 {
+      affine.store %c1, %A[%j, %i] : memref<2048x2048xf64>
+    }
+    affine.for %j = 1024 to 2048 {
+      affine.for %k = 0 to 256 {
+        affine.store %c0, %A[%j, %k] : memref<2048x2048xf64>
+      }
+      affine.for %k = 256 to 512 {
+        affine.store %c1, %A[%k, %i] : memref<2048x2048xf64>
+      }
+    }
+  }
+  return
+}
+// CHECK:       affine.for %[[IV0:.*]] = 0 to 1024 {
+// CHECK-NEXT:      affine.for %[[IV1:.*]] = 0 to 2048 {
+// CHECK-NEXT:        affine.store %{{.*}}, %{{.*}}[%[[IV0]], %[[IV1]]] : memref<2048x2048xf64>
+// CHECK-NEXT:      }
+// CHECK-NEXT:    }
+// CHECK-NEXT:    affine.for %[[IV0:.*]] = 1024 to 2048 {
+// CHECK-NEXT:      affine.for %[[IV1:.*]] = 0 to 2048 {
+// CHECK-NEXT:        affine.store %{{.*}}, %{{.*}}[%[[IV0]], %[[IV1]]] : memref<2048x2048xf64>
+// CHECK-NEXT:      }
+// CHECK-NEXT:    }
+// CHECK-NEXT:    affine.for %[[IV0:.*]] = 1024 to 2048 {
+// CHECK-NEXT:      affine.for %[[IV1:.*]] = 0 to 256 {
+// CHECK-NEXT:        affine.for %[[IV2:.*]] = 0 to 2048 {
+// CHECK-NEXT:          affine.store %{{.*}}, %{{.*}}[%[[IV0]], %[[IV1]]] : memref<2048x2048xf64>
+// CHECK-NEXT:        }
+// CHECK-NEXT:      }
+// CHECK-NEXT:    }
+// CHECK-NEXT:    affine.for %[[IV0:.*]] = 256 to 512 {
+// CHECK-NEXT:      affine.for %[[IV1:.*]] = 0 to 2048 {
+// CHECK-NEXT:        affine.for %[[IV2:.*]] = 1024 to 2048 {
+// CHECK-NEXT:          affine.store %{{.*}}, %{{.*}}[%[[IV0]], %[[IV1]]] : memref<2048x2048xf64>
+// CHECK-NEXT:        }
+// CHECK-NEXT:      }
+// CHECK-NEXT:    }
+
+// -----
+
+func @multi_nest_seq(%A: memref<2048x2048xf64>, %B: memref<2048x2048xf64>, %C: memref<2048x2048xf64>) {
+  affine.for %i = 0 to 2048 {
+    affine.for %j = 0 to 2048 {
+      %v = affine.load %A[%j, %i] : memref<2048x2048xf64>
+      affine.store %v, %A[%j, %i] : memref<2048x2048xf64>
+    }
+  }
+  // CHECK:     affine.for %[[IV0:.*]] = 0 to 2048 {
+  // CHECK-NEXT: affine.for %[[IV1:.*]] = 0 to 2048 {
+  // CHECK-NEXT:    %{{.*}} = affine.load %{{.*}}[%[[IV0]], %[[IV1]]] : memref<2048x2048xf64>
+  // CHECK-NEXT:    affine.store %{{.*}}, %{{.*}}[%[[IV0]], %[[IV1]]] : memref<2048x2048xf64>
+  // CHECK-NEXT:  }
+  // CHECK-NEXT:}
+
+  affine.for %i = 0 to 2048 {
+    affine.for %j = 0 to 2048 {
+      affine.for %k = 0 to 2048 {
+        %a = affine.load %A[%i, %k] : memref<2048x2048xf64>
+        %b = affine.load %B[%k, %j] : memref<2048x2048xf64>
+        %ci = affine.load %C[%i, %j] : memref<2048x2048xf64>
```

Comment:
All test cases have a constant upper bound. Please change some to use symbolic upper bounds.